tests/test_run_reference_vmm.py

```diff
@@ -94,32 +93,66 @@ def shutdown(vmm_process):
     vmm_process.wait(timeout=3)
 
 
-def expect_string(vmm_process, expected_string):
-    while True:
-        nextline = vmm_process.stdout.readline()
-        if expected_string in nextline.decode():
-            break
-    # If the expected string is identified and the loop breaks, return the
-    # entire line from stdout
-    return nextline.decode()
+def setup_stdout_nonblocking(vmm_process):
+
+    # We'll need to do non-blocking I/O with the underlying sub-process since
+    # we cannot use `communicate`, because `communicate` would close the
+    # `stdin` that we later want to use to `shutdown`, to do that by hand,
+    # we set `vmm_process.stdout` to non-blocking
+    # Then we can use `os.read` that would raise `BlockingIOError`
+
+    # FIXME: This should NOT be required to be done on every call, do it when we
+    #        'Class'ify the test case
+    flags = fcntl.fcntl(vmm_process.stdout, fcntl.F_GETFL)
+    fcntl.fcntl(vmm_process.stdout, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+
+
+def expect_string(vmm_process, expected_string, timeout=TEST_TIMEOUT):
+
+    setup_stdout_nonblocking(vmm_process)
+
+    # No. of seconds after which we'll give up
+    giveup_after = timeout
+    then = time.time()
+    found = False
+    # This is required because the pattern we are expecting might get split across two reads
+    all_data = bytes()
+    while not found:
+        try:
+            data = os.read(vmm_process.stdout.fileno(), 4096)
+            all_data += data
+            for line in all_data.split(b'\r\n'):
+                if expected_string in line.decode():
+                    found = True
+                    return line.decode()
+            # Whatever remains is collected in `all_data`.
+            all_data = line
+        except BlockingIOError as _:
+            # Raised on `EWOULDBLOCK`, so it's better to wait for sometime before retrying
+            time.sleep(1)
+            now = time.time()
+            if now - then > giveup_after:
+            if now - then > giveup_after :
```

Comment:
this should be removed

Reply:
yeah.. sorry!!! some goof up during rebasing! :-( Yikes!